// @salesforce/source-deploy-retrieve/lib/src/client/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolingDeployStatus = exports.RequestStatus = exports.ComponentStatus = void 0;
var ComponentStatus;
(function (ComponentStatus) {
    ComponentStatus["Created"] = "Created";
    ComponentStatus["Changed"] = "Changed";
    ComponentStatus["Unchanged"] = "Unchanged";
    ComponentStatus["Deleted"] = "Deleted";
    ComponentStatus["Failed"] = "Failed";
})(ComponentStatus = exports.ComponentStatus || (exports.ComponentStatus = {}));
var RequestStatus;
(function (RequestStatus) {
    RequestStatus["Pending"] = "Pending";
    RequestStatus["InProgress"] = "InProgress";
    RequestStatus["Succeeded"] = "Succeeded";
    RequestStatus["SucceededPartial"] = "SucceededPartial";
    RequestStatus["Failed"] = "Failed";
    RequestStatus["Canceling"] = "Canceling";
    RequestStatus["Canceled"] = "Canceled";
})(RequestStatus = exports.RequestStatus || (exports.RequestStatus = {}));
var ManageableState;
(function (ManageableState) {
    ManageableState["Beta"] = "beta";
    ManageableState["Deleted"] = "deleted";
    ManageableState["Deprecated"] = "deprecated";
    ManageableState["DeprecatedEditable"] = "deprecatedEditable";
    ManageableState["Installed"] = "installed";
    ManageableState["InstalledEditable"] = "installedEditable";
    ManageableState["Released"] = "released";
    ManageableState["Unmanaged"] = "unmanaged";
})(ManageableState || (ManageableState = {}));
var ToolingDeployStatus;
(function (ToolingDeployStatus) {
    // ContainerAsyncRequest states
    ToolingDeployStatus["Queued"] = "Queued";
    ToolingDeployStatus["Invalidated"] = "Invalidated";
    ToolingDeployStatus["Error"] = "Error";
    ToolingDeployStatus["Aborted"] = "Aborted";
    // Shared
    ToolingDeployStatus["Completed"] = "Completed";
    ToolingDeployStatus["Failed"] = "Failed";
    // unique to bundle requests
    ToolingDeployStatus["CompletedPartial"] = "CompletedPartial";
})(ToolingDeployStatus = exports.ToolingDeployStatus || (exports.ToolingDeployStatus = {}));
