// @salesforce/lightning-lsp-common/lib/html-language-service/services/htmlLinks.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const htmlScanner_1 = require("../parser/htmlScanner");
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const strings = __importStar(require("../utils/strings"));
const vscode_uri_1 = require("vscode-uri");
const htmlLanguageTypes_1 = require("../htmlLanguageTypes");
function normalizeRef(url, languageId) {
    let first = url[0];
    let last = url[url.length - 1];
    if (first === last && (first === "'" || first === '"')) {
        url = url.substr(1, url.length - 2);
    }
    if (languageId === 'razor' && url[0] === '~') {
        url = url.substr(1);
    }
    return url;
}
function validateRef(url, languageId) {
    if (!url.length) {
        return false;
    }
    if (languageId === 'handlebars' && /{{.*}}/.test(url)) {
        return false;
    }
    if (languageId === 'razor' && /@/.test(url)) {
        return false;
    }
    try {
        return !!vscode_uri_1.URI.parse(url);
    }
    catch (e) {
        return false;
    }
}
function getWorkspaceUrl(documentUri, tokenContent, documentContext, base) {
    if (/^\s*javascript\:/i.test(tokenContent) || /^\s*\#/i.test(tokenContent) || /[\n\r]/.test(tokenContent)) {
        return null;
    }
    tokenContent = tokenContent.replace(/^\s*/g, '');
    if (/^https?:\/\//i.test(tokenContent) || /^file:\/\//i.test(tokenContent)) {
        // Absolute link that needs no treatment
        return tokenContent;
    }
    if (/^\/\//i.test(tokenContent)) {
        // Absolute link (that does not name the protocol)
        let pickedScheme = strings.startsWith(documentUri, 'https://') ? 'https' : 'http';
        return pickedScheme + ':' + tokenContent.replace(/^\s*/g, '');
    }
    if (documentContext) {
        return documentContext.resolveReference(tokenContent, base || documentUri);
    }
    return tokenContent;
}
function createLink(document, documentContext, attributeValue, startOffset, endOffset, base) {
    let tokenContent = normalizeRef(attributeValue, document.languageId);
    if (!validateRef(tokenContent, document.languageId)) {
        return null;
    }
    if (tokenContent.length < attributeValue.length) {
        startOffset++;
        endOffset--;
    }
    let workspaceUrl = getWorkspaceUrl(document.uri, tokenContent, documentContext, base);
    if (!workspaceUrl || !isValidURI(workspaceUrl)) {
        return null;
    }
    return {
        range: vscode_languageserver_types_1.Range.create(document.positionAt(startOffset), document.positionAt(endOffset)),
        target: workspaceUrl,
    };
}
function isValidURI(uri) {
    try {
        vscode_uri_1.URI.parse(uri);
        return true;
    }
    catch (e) {
        return false;
    }
}
function findDocumentLinks(document, documentContext) {
    let newLinks = [];
    let rootAbsoluteUrl = null;
    let scanner = htmlScanner_1.createScanner(document.getText(), 0);
    let token = scanner.scan();
    let afterHrefOrSrc = false;
    let afterBase = false;
    let base = void 0;
    while (token !== htmlLanguageTypes_1.TokenType.EOS) {
        switch (token) {
            case htmlLanguageTypes_1.TokenType.StartTag:
                if (!base) {
                    let tagName = scanner.getTokenText().toLowerCase();
                    afterBase = tagName === 'base';
                }
                break;
            case htmlLanguageTypes_1.TokenType.AttributeName:
                let attributeName = scanner.getTokenText().toLowerCase();
                afterHrefOrSrc = attributeName === 'src' || attributeName === 'href';
                break;
            case htmlLanguageTypes_1.TokenType.AttributeValue:
                if (afterHrefOrSrc) {
                    let attributeValue = scanner.getTokenText();
                    if (!afterBase) {
                        // don't highlight the base link itself
                        let link = createLink(document, documentContext, attributeValue, scanner.getTokenOffset(), scanner.getTokenEnd(), base);
                        if (link) {
                            newLinks.push(link);
                        }
                    }
                    if (afterBase && typeof base === 'undefined') {
                        base = normalizeRef(attributeValue, document.languageId);
                        if (base && documentContext) {
                            base = documentContext.resolveReference(base, document.uri);
                        }
                    }
                    afterBase = false;
                    afterHrefOrSrc = false;
                }
                break;
        }
        token = scanner.scan();
    }
    return newLinks;
}
exports.findDocumentLinks = findDocumentLinks;
