// @lwc/compiler/dist/commonjs/bundler/meta-collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MetadataCollector {
    constructor() {
        this.decorators = [];
        this.importLocations = [];
        this.classMembers = [];
        this.exports = [];
    }
    collectDecorator(decorator) {
        this.decorators.push(decorator);
    }
    collectImportLocations(importLocations) {
        this.importLocations.push(...importLocations);
    }
    collectClassMember(classMember) {
        this.classMembers.push(classMember);
    }
    setDeclarationLoc(declarationLoc) {
        this.declarationLoc = declarationLoc;
    }
    setDoc(doc) {
        this.doc = doc;
    }
    collectExports(exports) {
        this.exports.push(...exports);
    }
    collectExperimentalTemplateDependencies(templatePath, templateDependencies) {
        if (!this.experimentalTemplateDependencies) {
            this.experimentalTemplateDependencies = [];
        }
        this.experimentalTemplateDependencies.push({
            templatePath,
            moduleDependencies: templateDependencies
        });
    }
    getMetadata() {
        return {
            decorators: this.decorators,
            importLocations: this.importLocations,
            classMembers: this.classMembers,
            declarationLoc: this.declarationLoc,
            doc: this.doc,
            experimentalTemplateDependencies: this.experimentalTemplateDependencies,
            exports: this.exports
        };
    }
}
exports.MetadataCollector = MetadataCollector;
