// @salesforce/lightning-lsp-common/lib/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils = __importStar(require("./utils"));
exports.utils = utils;
const context_1 = require("./context");
exports.WorkspaceContext = context_1.WorkspaceContext;
const shared = __importStar(require("./shared"));
exports.shared = shared;
const tagInfo_1 = require("./indexer/tagInfo");
exports.TagInfo = tagInfo_1.TagInfo;
const attributeInfo_1 = require("./indexer/attributeInfo");
exports.AttributeInfo = attributeInfo_1.AttributeInfo;
exports.Decorator = attributeInfo_1.Decorator;
exports.MemberType = attributeInfo_1.MemberType;
const logger_1 = require("./logger");
exports.interceptConsoleLogger = logger_1.interceptConsoleLogger;
const componentUtil = __importStar(require("./component-util"));
exports.componentUtil = componentUtil;
const htmlLanguageService_1 = require("./html-language-service/htmlLanguageService");
exports.getLanguageService = htmlLanguageService_1.getLanguageService;
const htmlParser_1 = require("./html-language-service/parser/htmlParser");
exports.parse = htmlParser_1.parse;
exports.Node = htmlParser_1.Node;
