// @salesforce/salesforcedx-apex-debugger/out/src/constants.js
"use strict";
/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEBUGGER_LAUNCH_TYPE = 'launch';
exports.DEBUGGER_TYPE = 'apex';
exports.LIVESHARE_DEBUGGER_TYPE = 'vslsShare';
exports.DEFAULT_STREAMING_TIMEOUT_MS = 14400;
exports.DEFAULT_LOCK_TIMEOUT_MS = 10000;
exports.DEFAULT_IDLE_TIMEOUT_MS = 1200000;
exports.DEFAULT_IDLE_WARN1_MS = 600000;
exports.DEFAULT_IDLE_WARN2_MS = 900000;
exports.DEFAULT_IDLE_WARN3_MS = 1080000;
exports.SHOW_MESSAGE_EVENT = 'showMessage';
exports.HOTSWAP_REQUEST = 'hotswap';
exports.EXCEPTION_BREAKPOINT_REQUEST = 'exceptionBreakpoint';
exports.LIST_EXCEPTION_BREAKPOINTS_REQUEST = 'listExceptionBreakpoints';
exports.LIVESHARE_DEBUG_TYPE_REQUEST = 'debugType';
exports.EXCEPTION_BREAKPOINT_BREAK_MODE_ALWAYS = 'always';
exports.EXCEPTION_BREAKPOINT_BREAK_MODE_NEVER = 'never';
exports.SALESFORCE_EXCEPTION_PREFIX = 'com/salesforce/api/exception/';
exports.TRIGGER_EXCEPTION_PREFIX = '__sfdc_trigger/';
