// @salesforce/salesforcedx-visualforce-language-server/out/src/modes/languageModes.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See OSSREADME.json in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const salesforcedx_visualforce_markup_language_server_1 = require("@salesforce/salesforcedx-visualforce-markup-language-server");
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
exports.ColorInformation = vscode_languageserver_protocol_1.ColorInformation;
exports.ColorPresentation = vscode_languageserver_protocol_1.ColorPresentation;
const languageModelCache_1 = require("../languageModelCache");
const cssMode_1 = require("./cssMode");
const embeddedSupport_1 = require("./embeddedSupport");
const htmlMode_1 = require("./htmlMode");
const javascriptMode_1 = require("./javascriptMode");
function getLanguageModes(supportedLanguages) {
    const htmlLanguageService = salesforcedx_visualforce_markup_language_server_1.getLanguageService();
    const documentRegions = languageModelCache_1.getLanguageModelCache(10, 60, document => embeddedSupport_1.getDocumentRegions(htmlLanguageService, document));
    let modelCaches = [];
    modelCaches.push(documentRegions);
    let modes = {};
    modes['html'] = htmlMode_1.getHTMLMode(htmlLanguageService);
    if (supportedLanguages['css']) {
        modes['css'] = cssMode_1.getCSSMode(documentRegions);
    }
    if (supportedLanguages['javascript']) {
        modes['javascript'] = javascriptMode_1.getJavascriptMode(documentRegions);
    }
    return {
        getModeAtPosition(document, position) {
            const languageId = documentRegions
                .get(document)
                .getLanguageAtPosition(position);
            if (languageId) {
                return modes[languageId];
            }
            return null;
        },
        getModesInRange(document, range) {
            return documentRegions
                .get(document)
                .getLanguageRanges(range)
                .map(r => {
                return {
                    start: r.start,
                    end: r.end,
                    mode: modes[r.languageId],
                    attributeValue: r.attributeValue
                };
            });
        },
        getAllModesInDocument(document) {
            const result = [];
            for (const languageId of documentRegions
                .get(document)
                .getLanguagesInDocument()) {
                const mode = modes[languageId];
                if (mode) {
                    result.push(mode);
                }
            }
            return result;
        },
        getAllModes() {
            const result = [];
            for (const languageId in modes) {
                const mode = modes[languageId];
                if (mode) {
                    result.push(mode);
                }
            }
            return result;
        },
        getMode(languageId) {
            return modes[languageId];
        },
        onDocumentRemoved(document) {
            modelCaches.forEach(mc => mc.onDocumentRemoved(document));
            for (const mode in modes) {
                modes[mode].onDocumentRemoved(document);
            }
        },
        dispose() {
            modelCaches.forEach(mc => mc.dispose());
            modelCaches = [];
            for (const mode in modes) {
                modes[mode].dispose();
            }
            modes = {};
        }
    };
}
exports.getLanguageModes = getLanguageModes;
