// @salesforce/soql-model/lib/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitMultiInputValues = exports.ValidatorFactory = exports.Validator = exports.SelectAnalyzer = exports.ModelSerializer = exports.ModelDeserializer = exports.SoqlModelUtils = exports.Impl = exports.Soql = void 0;
const Impl = require("./model/impl");
exports.Impl = Impl;
const Soql = require("./model/model");
exports.Soql = Soql;
const util_1 = require("./model/util");
Object.defineProperty(exports, "SoqlModelUtils", { enumerable: true, get: function () { return util_1.SoqlModelUtils; } });
const deserializer_1 = require("./serialization/deserializer");
Object.defineProperty(exports, "ModelDeserializer", { enumerable: true, get: function () { return deserializer_1.ModelDeserializer; } });
const serializer_1 = require("./serialization/serializer");
Object.defineProperty(exports, "ModelSerializer", { enumerable: true, get: function () { return serializer_1.ModelSerializer; } });
var selectAnalyzer_1 = require("./analyzers/selectAnalyzer");
Object.defineProperty(exports, "SelectAnalyzer", { enumerable: true, get: function () { return selectAnalyzer_1.SelectAnalyzer; } });
var validators_1 = require("./validators");
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return validators_1.Validator; } });
Object.defineProperty(exports, "ValidatorFactory", { enumerable: true, get: function () { return validators_1.ValidatorFactory; } });
Object.defineProperty(exports, "splitMultiInputValues", { enumerable: true, get: function () { return validators_1.splitMultiInputValues; } });
