// @salesforce/lightning-lsp-common/lib/html-language-service/parser/htmlParser.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const htmlScanner_1 = require("./htmlScanner");
const arrays_1 = require("../utils/arrays");
const htmlTags_1 = require("./htmlTags");
const htmlLanguageTypes_1 = require("../htmlLanguageTypes");
class Node {
    constructor(start, end, children, parent) {
        this.start = start;
        this.end = end;
        this.children = children;
        this.parent = parent;
        this.closed = false;
    }
    get attributeNames() {
        return this.attributes ? Object.keys(this.attributes) : [];
    }
    isSameTag(tagInLowerCase) {
        return this.tag && tagInLowerCase && this.tag.length === tagInLowerCase.length && this.tag.toLowerCase() === tagInLowerCase;
    }
    get firstChild() {
        return this.children[0];
    }
    get lastChild() {
        return this.children.length ? this.children[this.children.length - 1] : void 0;
    }
    findNodeBefore(offset) {
        let idx = arrays_1.findFirst(this.children, c => offset <= c.start) - 1;
        if (idx >= 0) {
            let child = this.children[idx];
            if (offset > child.start) {
                if (offset < child.end) {
                    return child.findNodeBefore(offset);
                }
                let lastChild = child.lastChild;
                if (lastChild && lastChild.end === child.end) {
                    return child.findNodeBefore(offset);
                }
                return child;
            }
        }
        return this;
    }
    findNodeAt(offset) {
        let idx = arrays_1.findFirst(this.children, c => offset <= c.start) - 1;
        if (idx >= 0) {
            let child = this.children[idx];
            if (offset > child.start && offset <= child.end) {
                return child.findNodeAt(offset);
            }
        }
        return this;
    }
}
exports.Node = Node;
function parse(text) {
    let scanner = htmlScanner_1.createScanner(text);
    let htmlDocument = new Node(0, text.length, [], void 0);
    let curr = htmlDocument;
    let endTagStart = -1;
    let endTagName = null;
    let pendingAttribute = null;
    let token = scanner.scan();
    while (token !== htmlLanguageTypes_1.TokenType.EOS) {
        switch (token) {
            case htmlLanguageTypes_1.TokenType.StartTagOpen:
                let child = new Node(scanner.getTokenOffset(), text.length, [], curr);
                curr.children.push(child);
                curr = child;
                break;
            case htmlLanguageTypes_1.TokenType.StartTag:
                curr.tag = scanner.getTokenText();
                break;
            case htmlLanguageTypes_1.TokenType.StartTagClose:
                curr.end = scanner.getTokenEnd(); // might be later set to end tag position
                if (curr.tag && htmlTags_1.isEmptyElement(curr.tag) && curr.parent) {
                    curr.closed = true;
                    curr = curr.parent;
                }
                break;
            case htmlLanguageTypes_1.TokenType.StartTagSelfClose:
                if (curr.parent) {
                    curr.closed = true;
                    curr.end = scanner.getTokenEnd();
                    curr = curr.parent;
                }
                break;
            case htmlLanguageTypes_1.TokenType.EndTagOpen:
                endTagStart = scanner.getTokenOffset();
                endTagName = null;
                break;
            case htmlLanguageTypes_1.TokenType.EndTag:
                endTagName = scanner.getTokenText().toLowerCase();
                break;
            case htmlLanguageTypes_1.TokenType.EndTagClose:
                if (endTagName) {
                    let node = curr;
                    // see if we can find a matching tag
                    while (!node.isSameTag(endTagName) && node.parent) {
                        node = node.parent;
                    }
                    if (node.parent) {
                        while (curr !== node) {
                            curr.end = endTagStart;
                            curr.closed = false;
                            curr = curr.parent;
                        }
                        curr.closed = true;
                        curr.endTagStart = endTagStart;
                        curr.end = scanner.getTokenEnd();
                        curr = curr.parent;
                    }
                }
                break;
            case htmlLanguageTypes_1.TokenType.AttributeName: {
                let attributeName = (pendingAttribute = scanner.getTokenText());
                let attributes = curr.attributes;
                if (!attributes) {
                    curr.attributes = attributes = {};
                }
                attributes[pendingAttribute] = null; // Support valueless attributes such as 'checked'
                break;
            }
            case htmlLanguageTypes_1.TokenType.AttributeValue: {
                let value = scanner.getTokenText();
                let attributes = curr.attributes;
                if (attributes && pendingAttribute) {
                    attributes[pendingAttribute] = value;
                    pendingAttribute = null;
                }
                break;
            }
        }
        token = scanner.scan();
    }
    while (curr.parent) {
        curr.end = text.length;
        curr.closed = false;
        curr = curr.parent;
    }
    return {
        roots: htmlDocument.children,
        findNodeBefore: htmlDocument.findNodeBefore.bind(htmlDocument),
        findNodeAt: htmlDocument.findNodeAt.bind(htmlDocument),
    };
}
exports.parse = parse;
