// @salesforce/salesforcedx-apex-replay-debugger/out/src/core/heapDump.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ApexHeapDump {
    constructor(heapDumpId, className, namespace, line) {
        /* tslint:disable */
        this.toString = () => {
            return `HeapDumpId: ${this.heapDumpId}, ClassName: ${this.className}, Namespace: ${this.namespace}, Line: ${this.line}`;
        }; /* This semi-colon is the reason for the tslint, formatting keeps adding it */
        this.heapDumpId = heapDumpId;
        this.className = className;
        this.namespace = namespace;
        this.line = line;
    }
    getHeapDumpId() {
        return this.heapDumpId;
    }
    getClassName() {
        return this.className;
    }
    getNamespace() {
        return this.namespace;
    }
    getLine() {
        return this.line;
    }
    getOverlaySuccessResult() {
        return this.overlaySuccessResut;
    }
    setOverlaySuccessResult(overlaySuccessResult) {
        this.overlaySuccessResut = overlaySuccessResult;
    }
}
exports.ApexHeapDump = ApexHeapDump;
