// @salesforce/salesforcedx-apex-replay-debugger/out/src/constants.js
"use strict";
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADDRESS_PREFIX = '0x';
exports.CHECKPOINT = 'checkpoint';
exports.CHECKPOINTS_LOCK_STRING = 'checkpoints_lock';
exports.DEBUGGER_LAUNCH_TYPE = 'launch';
exports.DEBUGGER_TYPE = 'apex-replay';
exports.LIVESHARE_DEBUGGER_TYPE = 'vslsShare';
exports.EVENT_CODE_UNIT_STARTED = 'CODE_UNIT_STARTED';
exports.EVENT_CODE_UNIT_FINISHED = 'CODE_UNIT_FINISHED';
exports.EVENT_CONSTRUCTOR_ENTRY = 'CONSTRUCTOR_ENTRY';
exports.EVENT_CONSTRUCTOR_EXIT = 'CONSTRUCTOR_EXIT';
exports.EVENT_EXECUTE_ANONYMOUS = 'Execute Anonymous';
exports.EVENT_HEAP_DUMP = 'HEAP_DUMP';
exports.EVENT_METHOD_ENTRY = 'METHOD_ENTRY';
exports.EVENT_METHOD_EXIT = 'METHOD_EXIT';
exports.EVENT_STATEMENT_EXECUTE = 'STATEMENT_EXECUTE';
exports.EVENT_USER_DEBUG = 'USER_DEBUG';
exports.EVENT_VF_APEX_CALL_START = 'VF_APEX_CALL_START';
exports.EVENT_VF_APEX_CALL_END = 'VF_APEX_CALL_END';
exports.EVENT_VARIABLE_SCOPE_BEGIN = 'VARIABLE_SCOPE_BEGIN';
exports.EVENT_VARIABLE_ASSIGNMENT = 'VARIABLE_ASSIGNMENT';
exports.EXEC_ANON_SIGNATURE = 'execute_anonymous_apex';
exports.FIELD_INTEGRITY_EXCEPTION = 'FIELD_INTEGRITY_EXCEPTION';
exports.INVALID_CROSS_REFERENCE_KEY = 'INVALID_CROSS_REFERENCE_KEY';
exports.SEND_METRIC_ERROR_EVENT = 'sendMetricError';
exports.SEND_METRIC_LAUNCH_EVENT = 'sendMetricLaunch';
exports.LAST_OPENED_LOG_FOLDER_KEY = 'LAST_OPENED_LOG_FOLDER_KEY';
exports.LAST_OPENED_LOG_KEY = 'LAST_OPENED_LOG_KEY';
exports.LIVESHARE_DEBUG_TYPE_REQUEST = 'debugType';
exports.SOBJECTS_URL = 'services/data/v43.0/tooling/sobjects';
exports.COMPOSITE_BATCH_URL = 'services/data/v43.0/tooling/composite/batch';
exports.OVERLAY_ACTION_DELETE_URL = 'services/data/v43.0/tooling/sobjects/ApexExecutionOverlayAction/';
exports.QUERY_URL = 'services/data/v43.0/tooling/query';
exports.MAX_ALLOWED_CHECKPOINTS = 5;
exports.SFDC_TRIGGER = '__sfdc_trigger/';
exports.EXTENT_TRIGGER_PREFIX = 'Trigger.';
exports.LC_APEX_PRIMITIVE_BLOB = 'blob';
exports.LC_APEX_PRIMITIVE_BOOLEAN = 'boolean';
exports.LC_APEX_PRIMITIVE_DATE = 'date';
exports.LC_APEX_PRIMITIVE_DATETIME = 'datetime';
exports.LC_APEX_PRIMITIVE_DECIMAL = 'decimal';
exports.LC_APEX_PRIMITIVE_DOUBLE = 'double';
exports.LC_APEX_PRIMITIVE_ID = 'id';
exports.LC_APEX_PRIMITIVE_INTEGER = 'integer';
exports.LC_APEX_PRIMITIVE_LONG = 'long';
exports.LC_APEX_PRIMITIVE_STRING = 'string';
exports.LC_APEX_PRIMITIVE_TIME = 'time';
exports.APEX_PRIMITIVE_STRING = 'String';
exports.KEY_VALUE_PAIR = 'KeyValuePair';
exports.KEY_VALUE_PAIR_KEY = 'key';
exports.KEY_VALUE_PAIR_VALUE = 'value';
