// @salesforce/soql-model/lib/model/unmodeled.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.REASON_UNMODELED_INSEMIJOINCONDITION = exports.REASON_UNMODELED_INCOLONEXPRESSIONCONDITION = exports.REASON_UNMODELED_DISTANCECONDITION = exports.REASON_UNMODELED_CALCULATEDCONDITION = exports.REASON_UNMODELED_EMPTYCONDITION = exports.REASON_UNMODELED_COLONEXPRESSION = exports.REASON_UNMODELED_FUNCTIONREFERENCE = exports.REASON_UNMODELED_UPDATE = exports.REASON_UNMODELED_RECORDTRACKING = exports.REASON_UNMODELED_BIND = exports.REASON_UNMODELED_OFFSET = exports.REASON_UNMODELED_GROUPBY = exports.REASON_UNMODELED_WITH = exports.REASON_UNMODELED_COUNT = exports.REASON_UNMODELED_COMPLEXGROUP = exports.REASON_UNMODELED_SELECT = exports.REASON_UNMODELED_DISTANCE = exports.REASON_UNMODELED_TYPEOF = exports.REASON_UNMODELED_SEMIJOIN = exports.REASON_UNMODELED_ALIAS = exports.REASON_UNMODELED_USING = exports.REASON_UNMODELED_AS = void 0;
const messages_1 = require("../messages/messages");
exports.REASON_UNMODELED_AS = {
    reasonCode: 'unmodeled:as',
    message: messages_1.Messages.unmodeled_as,
};
exports.REASON_UNMODELED_USING = {
    reasonCode: 'unmodeled:using',
    message: messages_1.Messages.unmodeled_using,
};
exports.REASON_UNMODELED_ALIAS = {
    reasonCode: 'unmodeled:alias',
    message: messages_1.Messages.unmodeled_alias,
};
exports.REASON_UNMODELED_SEMIJOIN = {
    reasonCode: 'unmodeled:semi-join',
    message: messages_1.Messages.unmodeled_semijoin,
};
exports.REASON_UNMODELED_TYPEOF = {
    reasonCode: 'unmodeled:type-of',
    message: messages_1.Messages.unmodeled_typeof,
};
exports.REASON_UNMODELED_DISTANCE = {
    reasonCode: 'unmodeled:distance',
    message: messages_1.Messages.unmodeled_distance,
};
exports.REASON_UNMODELED_SELECT = {
    reasonCode: 'unmodeled:select',
    message: messages_1.Messages.unmodeled_select,
};
exports.REASON_UNMODELED_COMPLEXGROUP = {
    reasonCode: 'unmodeled:complex-group',
    message: messages_1.Messages.unmodeled_complexgroup,
};
exports.REASON_UNMODELED_COUNT = {
    reasonCode: 'unmodeled:count',
    message: messages_1.Messages.unmodeled_count,
};
exports.REASON_UNMODELED_WITH = {
    reasonCode: 'unmodeled:with',
    message: messages_1.Messages.unmodeled_with,
};
exports.REASON_UNMODELED_GROUPBY = {
    reasonCode: 'unmodeled:group-by',
    message: messages_1.Messages.unmodeled_groupby,
};
exports.REASON_UNMODELED_OFFSET = {
    reasonCode: 'unmodeled:offset',
    message: messages_1.Messages.unmodeled_offset,
};
exports.REASON_UNMODELED_BIND = {
    reasonCode: 'unmodeled:bind',
    message: messages_1.Messages.unmodeled_bind,
};
exports.REASON_UNMODELED_RECORDTRACKING = {
    reasonCode: 'unmodeled:record-tracking',
    message: messages_1.Messages.unmodeled_recordtracking,
};
exports.REASON_UNMODELED_UPDATE = {
    reasonCode: 'unmodeled:update',
    message: messages_1.Messages.unmodeled_update,
};
exports.REASON_UNMODELED_FUNCTIONREFERENCE = {
    reasonCode: 'unmodeled:function-reference',
    message: messages_1.Messages.unmodeled_functionreference,
};
exports.REASON_UNMODELED_COLONEXPRESSION = {
    reasonCode: 'unmodeled:colon-expression',
    message: messages_1.Messages.unmodeled_colonexpression,
};
exports.REASON_UNMODELED_EMPTYCONDITION = {
    reasonCode: 'unmodeled:empty-condition',
    message: messages_1.Messages.unmodeled_emptycondition,
};
exports.REASON_UNMODELED_CALCULATEDCONDITION = {
    reasonCode: 'unmodeled:calculated-condition',
    message: messages_1.Messages.unmodeled_calculatedcondition,
};
exports.REASON_UNMODELED_DISTANCECONDITION = {
    reasonCode: 'unmodeled:distance-condition',
    message: messages_1.Messages.unmodeled_distancecondition,
};
exports.REASON_UNMODELED_INCOLONEXPRESSIONCONDITION = {
    reasonCode: 'unmodeled:in-colon-expression-condition',
    message: messages_1.Messages.unmodeled_incolonexpressioncondition,
};
exports.REASON_UNMODELED_INSEMIJOINCONDITION = {
    reasonCode: 'unmodeled:in-semi-join-condition',
    message: messages_1.Messages.unmodeled_insemijoincondition,
};
