// @salesforce/soql-language-server/lib/completion/soql-functions.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.soqlParametricDateRangeLiterals = exports.soqlDateRangeLiterals = exports.soqlOperators = exports.soqlFunctionsByName = exports.soqlFunctions = void 0;
exports.soqlFunctions = [
    {
        name: 'AVG',
        types: ['double', 'int', 'currency', 'percent'],
        isAggregate: true,
    },
    {
        name: 'COUNT',
        types: [
            'date',
            'datetime',
            'double',
            'int',
            'string',
            'combobox',
            'currency',
            'DataCategoryGroupReference',
            'email',
            'id',
            'masterrecord',
            'percent',
            'phone',
            'picklist',
            'reference',
            'textarea',
            'url',
        ],
        isAggregate: true,
    },
    {
        name: 'COUNT_DISTINCT',
        types: [
            'date',
            'datetime',
            'double',
            'int',
            'string',
            'combobox',
            'currency',
            'DataCategoryGroupReference',
            'email',
            'id',
            'masterrecord',
            'percent',
            'phone',
            'picklist',
            'reference',
            'textarea',
            'url',
        ],
        isAggregate: true,
    },
    {
        name: 'MAX',
        types: [
            'date',
            'datetime',
            'double',
            'int',
            'string',
            'time',
            'combobox',
            'currency',
            'DataCategoryGroupReference',
            'email',
            'id',
            'masterrecord',
            'percent',
            'phone',
            'picklist',
            'reference',
            'textarea',
            'url',
        ],
        isAggregate: true,
    },
    {
        name: 'MIN',
        types: [
            'date',
            'datetime',
            'double',
            'int',
            'string',
            'time',
            'combobox',
            'currency',
            'DataCategoryGroupReference',
            'email',
            'id',
            'masterrecord',
            'percent',
            'phone',
            'picklist',
            'reference',
            'textarea',
            'url',
        ],
        isAggregate: true,
    },
    {
        name: 'SUM',
        types: ['int', 'double', 'currency', 'percent'],
        isAggregate: true,
    },
];
exports.soqlFunctionsByName = exports.soqlFunctions.reduce((result, soqlFn) => {
    result[soqlFn.name] = soqlFn;
    return result;
}, {});
const typesForLTGTOperators = [
    'anyType',
    'complexvalue',
    'currency',
    'date',
    'datetime',
    'double',
    'int',
    'percent',
    'string',
    'textarea',
    'time',
    'url',
];
// SOQL operators semantics.
// Operators not listed here (i.e. equality operators) are allowed on all types
// and allow nulls
exports.soqlOperators = {
    '<': { types: typesForLTGTOperators, notNullable: true },
    '<=': { types: typesForLTGTOperators, notNullable: true },
    '>': { types: typesForLTGTOperators, notNullable: true },
    '>=': { types: typesForLTGTOperators, notNullable: true },
    INCLUDES: { types: ['multipicklist'], notNullable: true },
    EXCLUDES: { types: ['multipicklist'], notNullable: true },
    LIKE: { types: ['string', 'textarea', 'time'], notNullable: true },
};
exports.soqlDateRangeLiterals = [
    'YESTERDAY',
    'TODAY',
    'TOMORROW',
    'LAST_WEEK',
    'THIS_WEEK',
    'NEXT_WEEK',
    'LAST_MONTH',
    'THIS_MONTH',
    'NEXT_MONTH',
    'LAST_90_DAYS',
    'NEXT_90_DAYS',
    'THIS_QUARTER',
    'LAST_QUARTER',
    'NEXT_QUARTER',
    'THIS_YEAR',
    'LAST_YEAR',
    'NEXT_YEAR',
    'THIS_FISCAL_QUARTER',
    'LAST_FISCAL_QUARTER',
    'NEXT_FISCAL_QUARTER',
    'THIS_FISCAL_YEAR',
    'LAST_FISCAL_YEAR',
    'NEXT_FISCAL_YEAR',
];
exports.soqlParametricDateRangeLiterals = [
    'LAST_N_DAYS:n',
    'NEXT_N_DAYS:n',
    'NEXT_N_WEEKS:n',
    'LAST_N_WEEKS:n',
    'NEXT_N_MONTHS:n',
    'LAST_N_MONTHS:n',
    'NEXT_N_QUARTERS:n',
    'LAST_N_QUARTERS:n',
    'NEXT_N_YEARS:n',
    'LAST_N_YEARS:n',
    'NEXT_N_FISCAL_QUARTERS:n',
    'LAST_N_FISCAL_QUARTERS:n',
    'NEXT_N_FISCAL_YEARS:n',
    'LAST_N_FISCAL_YEARS:n',
];
