// @salesforce/lwc-language-server/lib/lwc-data-provider.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs-extra"));
const path_1 = require("path");
class LWCDataProvider {
    constructor(attributes) {
        this.activated = false;
        this.indexer = attributes.indexer;
        const standardData = fs.readFileSync(path_1.join(__dirname, 'resources/standard-lwc.json'), 'utf-8');
        const standardJson = JSON.parse(standardData);
        this._standardTags = standardJson.tags;
        this._globalAttributes = standardJson.globalAttributes;
    }
    getId() {
        return 'lwc';
    }
    isApplicable() {
        return this.activated;
    }
    provideTags() {
        const customTags = this.indexer.customData.map(tag => {
            return {
                name: tag.lwcName,
                description: tag.description,
                attributes: tag.attributes,
            };
        });
        return [...this._standardTags, ...customTags];
    }
    provideAttributes(tagName) {
        var _a;
        const tag = this.provideTags().find(t => t.name === tagName);
        return [...this._globalAttributes, ...(((_a = tag) === null || _a === void 0 ? void 0 : _a.attributes) || [])];
    }
    provideValues() {
        return [];
    }
}
exports.LWCDataProvider = LWCDataProvider;
