// @salesforce/lwc-language-server/lib/lwc-server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_languageserver_1 = require("vscode-languageserver");
const vscode_html_languageservice_1 = require("vscode-html-languageservice");
const compiler_1 = require("./javascript/compiler");
const aura_data_provider_1 = require("./aura-data-provider");
const lwc_data_provider_1 = require("./lwc-data-provider");
const lightning_lsp_common_1 = require("@salesforce/lightning-lsp-common");
const component_indexer_1 = __importDefault(require("./component-indexer"));
const typing_indexer_1 = __importDefault(require("./typing-indexer"));
const linter_1 = __importDefault(require("./template/linter"));
const vscode_uri_1 = require("vscode-uri");
exports.propertyRegex = new RegExp(/\{(?<property>\w+)\.*.*\}/);
exports.iteratorRegex = new RegExp(/iterator:(?<name>\w+)/);
var Token;
(function (Token) {
    Token["Tag"] = "tag";
    Token["AttributeKey"] = "attributeKey";
    Token["AttributeValue"] = "attributeValue";
    Token["DynamicAttributeValue"] = "dynamicAttributeValue";
    Token["Content"] = "content";
    Token["DynamicContent"] = "dynamicContent";
})(Token = exports.Token || (exports.Token = {}));
function findDynamicContent(text, offset) {
    const regex = new RegExp(/\{(?<property>\w+)\.*|\:*\w+\}/, 'g');
    let match = regex.exec(text);
    while (match && offset > match.index) {
        if (match.groups && match.groups.property && offset > match.index && regex.lastIndex > offset) {
            return match.groups.property;
        }
        match = regex.exec(text);
    }
    return null;
}
exports.findDynamicContent = findDynamicContent;
class Server {
    constructor() {
        this.connection = vscode_languageserver_1.createConnection();
        this.documents = new vscode_languageserver_1.TextDocuments();
        this.connection.onInitialize(this.onInitialize.bind(this));
        this.connection.onCompletion(this.onCompletion.bind(this));
        this.connection.onCompletionResolve(this.onCompletionResolve.bind(this));
        this.connection.onHover(this.onHover.bind(this));
        this.connection.onShutdown(this.onShutdown.bind(this));
        this.connection.onDefinition(this.onDefinition.bind(this));
        this.documents.listen(this.connection);
        this.documents.onDidChangeContent(this.onDidChangeContent.bind(this));
        this.documents.onDidSave(this.onDidSave.bind(this));
    }
    async onInitialize(params) {
        this.workspaceFolders = params.workspaceFolders;
        this.workspaceRoots = this.workspaceFolders.map(folder => vscode_uri_1.URI.parse(folder.uri).fsPath);
        this.context = new lightning_lsp_common_1.WorkspaceContext(this.workspaceRoots);
        this.componentIndexer = new component_indexer_1.default({ workspaceRoot: this.workspaceRoots[0] });
        this.lwcDataProvider = new lwc_data_provider_1.LWCDataProvider({ indexer: this.componentIndexer });
        this.auraDataProvider = new aura_data_provider_1.AuraDataProvider({ indexer: this.componentIndexer });
        this.typingIndexer = new typing_indexer_1.default({ workspaceRoot: this.workspaceRoots[0] });
        this.languageService = vscode_html_languageservice_1.getLanguageService({
            customDataProviders: [this.lwcDataProvider, this.auraDataProvider],
            useDefaultDataProvider: false,
        });
        this.context.configureProject();
        this.componentIndexer.init();
        this.typingIndexer.init();
        return this.capabilities;
    }
    get capabilities() {
        return {
            capabilities: {
                textDocumentSync: this.documents.syncKind,
                completionProvider: {
                    resolveProvider: true,
                    triggerCharacters: ['.', '-', '_', '<', '"', '=', '/', '>'],
                },
                hoverProvider: true,
                definitionProvider: true,
                workspace: {
                    workspaceFolders: {
                        supported: true,
                    },
                },
            },
        };
    }
    async onCompletion(params) {
        const { position, textDocument: { uri }, } = params;
        const doc = this.documents.get(uri);
        const htmlDoc = this.languageService.parseHTMLDocument(doc);
        if (await this.context.isLWCTemplate(doc)) {
            this.auraDataProvider.activated = false; // provide completions for lwc components in an Aura template
            this.lwcDataProvider.activated = true;
        }
        else if (await this.context.isLWCJavascript(doc)) {
            const customTags = this.componentIndexer.customData.map(tag => {
                return {
                    label: tag.lwcTypingsName,
                    kind: vscode_html_languageservice_1.CompletionItemKind.Folder,
                };
            });
            return {
                isIncomplete: false,
                items: customTags,
            };
        }
        else if (await this.context.isAuraMarkup(doc)) {
            this.auraDataProvider.activated = true;
            this.lwcDataProvider.activated = false;
        }
        else {
            return;
        }
        return this.languageService.doComplete(doc, position, htmlDoc);
    }
    onCompletionResolve(item) {
        return item;
    }
    async onHover(params) {
        const { position, textDocument: { uri }, } = params;
        const doc = this.documents.get(uri);
        const htmlDoc = this.languageService.parseHTMLDocument(doc);
        if (await this.context.isLWCTemplate(doc)) {
            this.auraDataProvider.activated = false;
            this.lwcDataProvider.activated = true;
        }
        else if (await this.context.isAuraMarkup(doc)) {
            this.auraDataProvider.activated = true;
            this.lwcDataProvider.activated = false;
        }
        else {
            return;
        }
        return this.languageService.doHover(doc, position, htmlDoc);
    }
    async onDidChangeContent(changeEvent) {
        const { document } = changeEvent;
        const { uri } = document;
        if (await this.context.isLWCTemplate(document)) {
            const diagnostics = linter_1.default(document);
            this.connection.sendDiagnostics({ uri, diagnostics });
        }
        if (await this.context.isLWCJavascript(document)) {
            const { metadata, diagnostics } = await compiler_1.compileDocument(document);
            this.connection.sendDiagnostics({ uri, diagnostics });
            if (metadata) {
                const tag = this.componentIndexer.findTagByURI(uri);
                if (tag) {
                    tag.updateMetadata(metadata);
                }
            }
        }
    }
    async onDidSave(change) {
        const { document } = change;
        const { uri } = document;
        if (await this.context.isLWCJavascript(document)) {
            const doc = await compiler_1.compileDocument(document);
            const metadata = doc.metadata;
            if (metadata) {
                const tag = this.componentIndexer.findTagByURI(uri);
                if (tag) {
                    tag.updateMetadata(metadata);
                }
            }
        }
    }
    onShutdown() {
        this.componentIndexer.persistCustomComponents();
    }
    onExit() {
        this.componentIndexer.persistCustomComponents();
    }
    onDefinition(params) {
        var _a, _b, _c;
        const cursorInfo = this.cursorInfo(params);
        if (!cursorInfo) {
            return null;
        }
        const tag = this.componentIndexer.findTagByName(cursorInfo.tag);
        switch (cursorInfo.type) {
            case Token.Tag:
                return ((_a = tag) === null || _a === void 0 ? void 0 : _a.allLocations) || [];
            case Token.AttributeKey:
                const attr = (_b = tag) === null || _b === void 0 ? void 0 : _b.attribute(cursorInfo.name);
                if (attr) {
                    return [attr.location];
                }
            case Token.DynamicContent:
            case Token.DynamicAttributeValue:
                const { uri } = params.textDocument;
                if (cursorInfo.range) {
                    return [vscode_languageserver_1.Location.create(uri, cursorInfo.range)];
                }
                else {
                    const component = this.componentIndexer.findTagByURI(uri);
                    const location = (_c = component) === null || _c === void 0 ? void 0 : _c.classMemberLocation(cursorInfo.name);
                    if (location) {
                        return [location];
                    }
                }
        }
        return [];
    }
    cursorInfo({ textDocument: { uri }, position }, document) {
        var _a, _b;
        const doc = document || this.documents.get(uri);
        const offset = doc.offsetAt(position);
        const scanner = this.languageService.createScanner(doc.getText());
        let token;
        let tag;
        let attributeName;
        const iterators = [];
        do {
            token = scanner.scan();
            if (token === vscode_html_languageservice_1.TokenType.StartTag) {
                tag = scanner.getTokenText();
            }
            if (token === vscode_html_languageservice_1.TokenType.AttributeName) {
                attributeName = scanner.getTokenText();
                const iterator = exports.iteratorRegex.exec(attributeName);
                if (iterator) {
                    iterators.unshift({
                        name: iterator.groups.name,
                        range: {
                            start: doc.positionAt(scanner.getTokenOffset() + 9),
                            end: doc.positionAt(scanner.getTokenEnd()),
                        },
                    });
                }
            }
            if (token === vscode_html_languageservice_1.TokenType.AttributeValue && attributeName === 'for:item') {
                iterators.unshift({
                    name: scanner.getTokenText().replace(/"|'/g, ''),
                    range: {
                        start: doc.positionAt(scanner.getTokenOffset()),
                        end: doc.positionAt(scanner.getTokenEnd()),
                    },
                });
            }
        } while (token !== vscode_html_languageservice_1.TokenType.EOS && scanner.getTokenEnd() <= offset);
        const content = scanner.getTokenText();
        switch (token) {
            case vscode_html_languageservice_1.TokenType.StartTag:
            case vscode_html_languageservice_1.TokenType.EndTag: {
                return { type: Token.Tag, name: tag, tag };
            }
            case vscode_html_languageservice_1.TokenType.AttributeName: {
                return { type: Token.AttributeKey, tag, name: content };
            }
            case vscode_html_languageservice_1.TokenType.AttributeValue: {
                const match = exports.propertyRegex.exec(content);
                if (match) {
                    const item = iterators.find(i => i.name === match.groups.property) || null;
                    return {
                        type: Token.DynamicAttributeValue,
                        name: match.groups.property,
                        range: (_a = item) === null || _a === void 0 ? void 0 : _a.range,
                        tag,
                    };
                }
                else {
                    return { type: Token.AttributeValue, name: content, tag };
                }
            }
            case vscode_html_languageservice_1.TokenType.Content: {
                const relativeOffset = offset - scanner.getTokenOffset();
                const match = findDynamicContent(content, relativeOffset);
                if (match) {
                    const item = iterators.find(i => i.name === match) || null;
                    return {
                        type: Token.DynamicContent,
                        name: match,
                        range: (_b = item) === null || _b === void 0 ? void 0 : _b.range,
                        tag,
                    };
                }
                else {
                    return {
                        type: Token.Content,
                        tag,
                        name: content,
                    };
                }
            }
        }
        return null;
    }
    listen() {
        lightning_lsp_common_1.interceptConsoleLogger(this.connection);
        this.connection.listen();
    }
}
exports.default = Server;
