// @salesforce/salesforcedx-apex-debugger/out/src/commands/baseDebuggerCommand.js
"use strict";
/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const requestService_1 = require("@salesforce/salesforcedx-utils-vscode/out/src/requestService");
class BaseDebuggerCommand extends requestService_1.BaseCommand {
    constructor(commandName, debuggedRequestId, queryString, request) {
        super(queryString);
        this.debuggerApiPath = 'services/debug/v41.0';
        this.commandName = commandName;
        this.debuggedRequestId = debuggedRequestId;
        this.request = request;
    }
    getCommandUrl() {
        const urlElements = [
            this.debuggerApiPath,
            this.commandName,
            this.debuggedRequestId
        ];
        return urlElements.join('/');
    }
    getQueryString() {
        return this.queryString;
    }
    getRequest() {
        if (this.request) {
            return JSON.stringify(this.request);
        }
        else {
            return undefined;
        }
    }
}
exports.BaseDebuggerCommand = BaseDebuggerCommand;
